mono/utils/os-event.h

```diff
@@ -0,0 +1,62 @@
+
+#ifndef _MONO_UTILS_OS_EVENT_H_
+#define _MONO_UTILS_OS_EVENT_H_
+
+#include <config.h>
+#include <glib.h>
+
+#include "mono-os-mutex.h"
+
+#ifndef MONO_INFINITE_WAIT
+#define MONO_INFINITE_WAIT ((guint32) 0xFFFFFFFF)
+#endif
+
+#define MONO_OS_EVENT_WAIT_MAXIMUM_OBJECTS 64
+
+typedef enum {
+	MONO_OS_EVENT_WAIT_RET_SUCCESS_0 =  0,
+	MONO_OS_EVENT_WAIT_RET_ALERTED   = -1,
+	MONO_OS_EVENT_WAIT_RET_TIMEOUT   = -2,
+} MonoOSEventWaitRet;
+
+typedef struct _MonoOSEvent MonoOSEvent;
+
+typedef void (*MonoOSEventFreeCb) (MonoOSEvent*);
+
+struct _MonoOSEvent {
+#ifdef HOST_WIN32
+	gpointer handle;
+#else
+	mono_mutex_t mutex;
+	mono_cond_t cond;
+	gboolean manual;
+	gboolean signalled;
+	guint32 set_count;
+#endif
+
+	guint32 ref;
+	MonoOSEventFreeCb free_cb;
+};
+
+void
+mono_os_event_init (MonoOSEvent *event, gboolean manual, gboolean initial, MonoOSEventFreeCb free_cb);
```

Comment:
I think these functions could use some documentation in general. Unless you've done a lot of programming with the managed equivalents of this synchronization primitive, it's not exactly obvious what the parameters mean and what semantics the functions have given different init parameters.

It would also be nice to expose `mono_os_event_destroy` as some code that might want to use a `MonoOSEvent` doesn't necessarily want to deal with reference counting. In fact, if the lifetime of a `MonoOSEvent` is very straightforward and all you need is init/destroy, having to use ref/unref would really just make the code confusing (a reader would end up thinking that the event has a more complicated lifetime than it actually does).